File: problem-015/lattice_paths.rb
# Total Lattice paths using recursion + memoization

require 'memoize'
include Memoize

def lattice_paths(row, col)
  paths = 0

  if row > 0
    # recurse vertically
    paths += lattice_paths(row - 1, col)
  end

  if col > 0
    # recurse horizontally
    paths += lattice_paths(row, col - 1)
  end

  if row == 0 && col == 0
    # reached destination
    paths += 1
  end

  return paths
end

# make stuff fast
memoize :lattice_paths

p lattice_paths(20, 20)

File: problem-015/lattice_paths.js
// Total Lattice paths for a grid using
// recursion + memoization

var results = {};
function lattice_paths(x, y) {
  // memoize results
  if (results[[x, y]]) {
    return results[[x, y]];
  }

  var paths = 0;

  if (x > 0) {
    // recurse vertically
    paths += lattice_paths(x - 1, y);
  }
  if (y > 0) {
    // recurse horizontally
    paths += lattice_paths(x, y - 1);
  }

  if (x == 0 && y == 0) {
    // reached destination
    paths += 1
  }

  // memoize and return
  return results[[x, y]] = paths
}

console.log(lattice_paths(20, 20));
